### ISSUE TITLE
Blank strategy with ATR based stop loss and take profit.

### ISSUE THREAD

**Issue by Purrbox** [freqtrade/freqtrade#10373]
Hi,

I've been trying to implement ATR based stop loss and take profit and have tried multiple methods but am just unable to get this working reliably.

I'm new to coding so am only able to do really basic rule based entries on indicator values. 

I was hoping somebody could share a strategy file that has the code for atr take profit and stop loss included?

This is what I've come up with so far and I honestly am just confusing myself at this point. Any help would really be appreciated!

``` python
    def confirm_trade_entry(self, pair: str, order_type: str, amount: float, rate: float, 
                        time_in_force: str, current_time: datetime, entry_tag: str,
                        side: str, **kwargs) -> bool:
        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
        last_candle = dataframe.iloc[-1].squeeze()
        
        self.trade_atr_values[pair] = last_candle['atr']
        logger.info(f"Stored ATR for {pair}: {self.trade_atr_values[pair]}")
    
        return True

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        print(f"Populating exit trend for {metadata['pair']}")
        #dataframe.loc[
        #    (dataframe['rsi'] < 1),
        #    'exit_long'] = 1
        return dataframe

    def confirm_trade_entry(self, pair: str, order_type: str, amount: float, rate: float, 
                            time_in_force: str, current_time: datetime, entry_tag: str,
                            side: str, **kwargs) -> bool:
        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
        last_candle = dataframe.iloc[-1].squeeze()
        
        atr_value = last_candle['atr']
        #logger.info(f"Trade entry for {pair}: ATR = {atr_value}, Open rate = {rate}")
        
        return True

    def custom_stoploss(self, pair: str, trade: Trade, current_time: datetime,
                        current_rate: float, current_profit: float, **kwargs) -> float:
        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
        last_candle = dataframe.iloc[-1].squeeze()
        
        atr_value = last_candle['atr']
        atr_based_stoploss = -(atr_value * 2 / current_rate)

            # Store the atr_based_stoploss value in the dataframe
        last_candle_index = dataframe.index[-1]
        dataframe.at[last_candle_index, 'atr_based_stoploss'] = atr_based_stoploss
        
        #logger.info(f"Custom stoploss for {pair}: Current ATR = {atr_value}, "
        #            f"Stoploss = {atr_based_stoploss}, Current rate = {current_rate}")
        
        return atr_based_stoploss

    def custom_exit(self, pair: str, trade: Trade, current_time: datetime, current_rate: float,
                        current_profit: float, **kwargs):
            dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
            last_candle = dataframe.iloc[-1].squeeze()
            
            atr_value = last_candle['atr']
            atr_based_tp = atr_value * 2 / trade.open_rate

            print(f"ATR Based TP {atr_based_tp}")

                # Store the atr_based_tp value in dataframe
            last_candle_index = dataframe.index[-1]
            dataframe.at[last_candle_index, 'atr_based_tp'] = atr_based_tp
            
            #logger.info(f"Custom exit check for {pair}: Current ATR = {atr_value}, "
            #            f"Take profit threshold = {atr_based_tp}, Current profit = {current_profit}")
            
            if current_profit > atr_based_tp:
                logger.info(f"Taking profit for {pair}")
                return 'take_profit'
            
            return None

    def custom_exit_price(self, pair: str, trade: Trade,
                          current_time: datetime, proposed_rate: float,
                          current_profit: float, **kwargs) -> float:
        atr = self.trade_atr_values.get(pair)
        if atr:
            trailing_stop_rate = trade.open_rate * (1 + 0.02)  # 2% trailing stop
            return max(proposed_rate, trailing_stop_rate)
        return proposed_rate

    def custom_sell(self, pair: str, trade: Trade, current_time: datetime, current_rate: float,
                    current_profit: float, **kwargs):
        #remove the stored atr value when closing the trade
        self.trade_atr_values.pop(pair, None)
        return False
    
 
    ```

**Comment by froggleston** [comment]
1. There is no `custom_sell` function any more. `custom_exit` replaces it. You also have `confirm_trade_entry` twice which doesn't make sense.
2. As custom_exit is called every bot loop, you can calculate the ATR based TP every time and it'll be updated for each candle. If you want the ATR relative to trade open, there's an example of fixed risk:reward that may help: https://github.com/freqtrade/freqtrade-strategies/blob/main/user_data/strategies/FixedRiskRewardLoss.py
3. You cannot store elements in the dataframe from custom functions. To store data for a trade, use the [Persistent Storage](https://www.freqtrade.io/en/stable/strategy-advanced/#storing-information-persistent) solution instead.
4. You don't need to set the trailing rate in custom_exit_price. `custom_exit` will exit immediately once the TP level has been reached.
5. As you already have `last_candle` you can get its index directly - you don't have to call `dataframe.index[-1]` again (and indeed you shouldn't)

There's a good example of a working ATR based SL/TP [here](https://botacademy.ddns.net/2024/04/18/fsd-better-way-of-using-atr-for-exits/)
